File: packages/kmp-shell/ios-sdk/Sources/SelfSDK/SelfSdk.swift
// SPDX-FileCopyrightText: 2025 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.

import UIKit

/// Main entry point for the Self SDK on iOS.
///
/// Usage:
/// ```swift
/// let selfSdk = SelfSdk.configure {
///     $0.appId = "your-app-id"
///     $0.environment = .production
/// }
///
/// selfSdk.launch(
///     from: self,
///     request: .init(scope: "identity", userId: "user-123")
/// ) { result in
///     switch result {
///     case .verified(let verification):
///         print("Verified: \(verification.userId)")
///     case .failed(let error):
///         print("Failed: \(error.message)")
///     case .dismissed:
///         print("User dismissed")
///     }
/// }
/// ```
public class SelfSdk {

    private let config: SelfSdkConfig

    private init(config: SelfSdkConfig) {
        self.config = config
    }

    /// Configure and create a SelfSdk instance.
    public static func configure(_ block: (inout SelfSdkConfig) -> Void) -> SelfSdk {
        var config = SelfSdkConfig()
        block(&config)
        assert(!config.appId.isEmpty, "appId must not be empty")
        return SelfSdk(config: config)
    }

    /// Launch the Self verification flow.
    public func launch(
        from viewController: UIViewController,
        request: SelfVerificationRequest,
        completion: @escaping (SelfSdkResult) -> Void
    ) {
        let sdkVC = SelfSDKViewController()
        sdkVC.devMode = config.devMode
        if let devUrl = config.devServerUrl {
            sdkVC.devServerUrl = devUrl
        }
        sdkVC.modalPresentationStyle = .fullScreen

        // Register native bridge handlers
        sdkVC.registerHandler(SelfNfcBridge(router: NativeMessageRouter { js in
            DispatchQueue.main.async { /* JS eval handled by ViewController */ }
        }))
        sdkVC.registerHandler(BiometricBridge())
        sdkVC.registerHandler(SecureStorageBridge())
        sdkVC.registerHandler(CryptoBridge())

        // Wire result callbacks
        sdkVC.onResult = { result in
            completion(.verified(SelfVerification(
                userId: result.userId,
                verificationId: result.verificationId,
                proof: result.proof,
                claims: result.claims
            )))
        }
        sdkVC.onDismiss = {
            completion(.dismissed)
        }

        viewController.present(sdkVC, animated: true)
    }
}

// MARK: - Configuration

/// SDK configuration.
public struct SelfSdkConfig {
    public var appId: String = ""
    public var environment: SelfSdkEnvironment = .production
    public var theme: SelfSdkTheme = SelfSdkTheme()
    public var features: [String: Bool] = [:]
    public var locale: String?
    public var devMode: Bool = false
    public var devServerUrl: String?
}

/// SDK environment.
public enum SelfSdkEnvironment {
    case production
    case staging
    case development
}

/// Theme customization.
public struct SelfSdkTheme {
    public var primaryColor: String = "#000000"
    public var backgroundColor: String = "#FFFFFF"
    public var accentColor: String = "#FFFBEB"

    public init() {}
}

// MARK: - Request & Result

/// Verification request parameters.
public struct SelfVerificationRequest {
    public let scope: String
    public let userId: String?
    public let callbackUrl: String?
    public let metadata: [String: String]

    public init(
        scope: String,
        userId: String? = nil,
        callbackUrl: String? = nil,
        metadata: [String: String] = [:]
    ) {
        self.scope = scope
        self.userId = userId
        self.callbackUrl = callbackUrl
        self.metadata = metadata
    }
}

/// SDK result.
public enum SelfSdkResult {
    case verified(SelfVerification)
    case failed(SelfSdkError)
    case dismissed
}

/// Successful verification data.
public struct SelfVerification {
    public let userId: String?
    public let verificationId: String?
    public let proof: Any?
    public let claims: [String: Any]?
}

/// SDK error.
public struct SelfSdkError: Error {
    public let code: String
    public let message: String
}

File: packages/kmp-shell/android-sdk/src/main/kotlin/xyz/self/sdk/android/SelfSdk.kt
// SPDX-FileCopyrightText: 2025 Social Connect Labs, Inc.
// SPDX-License-Identifier: BUSL-1.1
// NOTE: Converts to Apache-2.0 on 2029-06-11 per LICENSE.

package xyz.self.sdk.android

import android.app.Activity
import android.content.Context
import android.content.Intent
import androidx.activity.result.ActivityResultLauncher
import androidx.fragment.app.FragmentActivity
import xyz.self.sdk.bridge.*

/**
 * Main entry point for the Self SDK.
 *
 * Host apps (MiniPay, etc.) use this class to:
 * 1. Configure the SDK with their app ID and settings
 * 2. Launch the verification flow
 * 3. Receive verification results via callback
 *
 * Usage:
 * ```kotlin
 * val selfSdk = SelfSdk.configure {
 *     appId = "your-app-id"
 *     environment = SelfSdkEnvironment.PRODUCTION
 * }
 *
 * selfSdk.launch(
 *     activity = this,
 *     request = VerificationRequest(scope = "identity", userId = "user-123"),
 *     callback = object : SelfSdkCallback {
 *         override fun onVerificationComplete(result: VerificationResult) { ... }
 *         override fun onVerificationFailed(error: SelfSdkError) { ... }
 *         override fun onDismissed() { ... }
 *     }
 * )
 * ```
 */
class SelfSdk private constructor(
    private val config: SelfSdkConfig,
) {
    companion object {
        /**
         * Configure and create a SelfSdk instance.
         */
        fun configure(block: SelfSdkConfig.Builder.() -> Unit): SelfSdk {
            val builder = SelfSdkConfig.Builder()
            block(builder)
            return SelfSdk(builder.build())
        }
    }

    /**
     * Launch the Self verification flow.
     *
     * This opens a new Activity with a WebView that runs the full
     * verification experience. Results are delivered via the callback.
     */
    fun launch(
        activity: FragmentActivity,
        request: VerificationRequest,
        callback: SelfSdkCallback,
    ) {
        // Store callback for result delivery
        SelfSdkCallbackHolder.callback = callback
        SelfSdkCallbackHolder.config = config
        SelfSdkCallbackHolder.request = request

        val intent = SelfVerificationActivity.createIntent(
            context = activity,
            config = config,
            request = request,
        )
        activity.startActivity(intent)
    }
}

/**
 * SDK configuration.
 */
data class SelfSdkConfig(
    val appId: String,
    val environment: SelfSdkEnvironment,
    val theme: SelfSdkTheme,
    val features: Map<String, Boolean>,
    val locale: String?,
    val devMode: Boolean,
    val devServerUrl: String?,
) {
    class Builder {
        var appId: String = ""
        var environment: SelfSdkEnvironment = SelfSdkEnvironment.PRODUCTION
        var theme: SelfSdkTheme = SelfSdkTheme()
        var features: Map<String, Boolean> = emptyMap()
        var locale: String? = null
        var devMode: Boolean = false
        var devServerUrl: String? = null

        fun build(): SelfSdkConfig {
            require(appId.isNotBlank()) { "appId must not be blank" }
            return SelfSdkConfig(
                appId = appId,
                environment = environment,
                theme = theme,
                features = features,
                locale = locale,
                devMode = devMode,
                devServerUrl = devServerUrl,
            )
        }
    }
}

/** SDK environment. */
enum class SelfSdkEnvironment {
    PRODUCTION,
    STAGING,
    DEVELOPMENT,
}

/** Theme customization. */
data class SelfSdkTheme(
    val primaryColor: String = "#000000",
    val backgroundColor: String = "#FFFFFF",
    val accentColor: String = "#FFFBEB",
)

/** Verification request parameters. */
data class VerificationRequest(
    val scope: String,
    val userId: String? = null,
    val callbackUrl: String? = null,
    val metadata: Map<String, String> = emptyMap(),
)

/** Result of a successful verification. */
data class VerificationResult(
    val userId: String?,
    val verificationId: String?,
    val proof: Any?,
    val claims: Map<String, Any?>,
)

/** SDK error. */
data class SelfSdkError(
    val code: String,
    val message: String,
)

/** Callback interface for host apps. */
interface SelfSdkCallback {
    fun onVerificationComplete(result: VerificationResult)
    fun onVerificationFailed(error: SelfSdkError)
    fun onDismissed()
}

/** Internal holder for callback across activity boundaries. */
internal object SelfSdkCallbackHolder {
    var callback: SelfSdkCallback? = null
    var config: SelfSdkConfig? = null
    var request: VerificationRequest? = null

    fun clear() {
        callback = null
        config = null
        request = null
    }
}
